addons/resources/src/ResourcePanel.js

```diff
@@ -0,0 +1,161 @@
+import { document, DOMParser } from 'global';
+import React, { Component } from 'react';
+import PropTypes from 'prop-types';
+import { Button } from '@storybook/components';
+import styled from '@emotion/styled';
+import AceEditor from 'react-ace';
+import 'brace/mode/javascript';
+import 'brace/theme/monokai';
+
+const storybookIframe = 'storybook-preview-iframe';
+const addedElAttr = 'addedbyaddon';
+
+const PanelWrapper = styled.div({
+  position: 'absolute',
+  top: '10px',
+  left: '10px',
+});
+
+export default class ResourcePanel extends Component {
+  constructor(props) {
+    super(props);
+    this.state = { resources: `` };
+    this.onAddResources = this.onAddResources.bind(this);
+    this.parser = new DOMParser();
+  }
+
+  componentDidMount() {
+    const { channel } = this.props;
+    this.iframe = document.getElementById(storybookIframe);
+    if (!this.iframe) {
+      throw new Error('Cannot find Storybook iframe');
+    }
+    channel.on('storybook/resources/add_resources', this.onAddResources);
+  }
+
+  componentWillUnmount() {
+    const { channel } = this.props;
+    channel.removeListener('storybook/resources/add_resources', this.onAddResources);
+  }
+
+  onChange(newValue) {
+    this.setState({ resources: newValue });
+  }
+
+  onAddResources(resources) {
+    this.setState({ resources });
+    this.loadResources();
+  }
+
+  apply() {
+    const { resources = '' } = this.state;
+    this.setState({ resources });
+    this.loadResources();
+  }
+
+  convertStringToDom(str) {
+    const parsedHtml = this.parser.parseFromString(str, 'text/html');
+    const elements = parsedHtml.querySelectorAll('head > *');
+
+    // remove null elements if any!
+    return [...elements].map(eL => {
+      let newEl;
+      if (eL) {
+        /** Tried multiple easier ways to convert the string to dOM objects
+         DOMParser, jQuery etc. However, adding <script> elements to the iframe
+        seems to fail using those methods. The only way that seemed to work for all
+        cases was to createElements in this way! */
+        newEl = document.createElement(eL.tagName.toLowerCase());
+        [...Array(eL.attributes.length).keys()].forEach(idx => {
+          const attr = eL.attributes.item(idx).name;
+          newEl[attr] = eL[attr];
+        });
+        while (eL.hasChildNodes()) {
+          newEl.appendChild(eL.removeChild(eL.firstChild));
+        }
+        newEl.setAttribute(addedElAttr, '');
+      }
+      return newEl;
+    });
+  }
+
+  loadResources() {
+    const { resources = `` } = this.state;
+
+    const addElements = (domElements, i) => {
```

Comment:
Looks like this function could be extracted, it doesn't use any inner state from the `loadResources` func